docs/include_package-core.rst

```diff
@@ -13,7 +13,9 @@ setProvider
 
 Will change the provider for its module.
 
-.. note:: When called on the umbrella package ``web3`` it will also set the provider for all sub modules ``web3.eth``, ``web3.shh``, etc EXCEPT ``web3.bzz`` which needs a separate provider at all times.
+.. note::
+    When called on the umbrella package ``web3`` it will also set the provider for all sub modules ``web3.eth``, ``web3.shh``, etc EXCEPT ``web3.bzz`` which needs a separate provider at all times.
+    Be aware to clean up the listeners of your provider if you pass the exact same provider instance a second time to the ``setProvider`` method of any module. This because the ``setProvider`` method will register the listeners required from web3.js again.
```

Comment:
> However, it's possible that setProvider should clear existing subscriptions since toggling networks in a Dapp is pretty common. Intuitively, going a --> b --> a would "just work".

We were always talking about EventEmitter listeners and not subscriptions. Subscriptions do get cleaned up as it was before. (please read the whole code and try to understand what changes I have done otherwise will this take years)

> Can we remove this docs addition and merge the PR into #3190 as is, leaving the question of what should be done here open? 

We can do that but could you next time please answer a comment more precisely? ([comment from me](https://github.com/ethereum/web3.js/pull/3303#discussion_r367886182))

>  am very aware of the issues raised by this test change (now).

There aren't really any bigger problems. We do solve a lot of bugs with it and we have extended the test suite as well. I think you are panicking without any reason. We also do have now an RC version we will release first which does help us to cover unseen problems which I can't see right now.